cli/fathom_web/commands/vectorize.py

```diff
@@ -0,0 +1,340 @@
+    server_thread.start()
+    print('Done')
+    return server, server_thread
+
+
+def configure_firefox(fathom_fox, fathom_trainees, output_directory, show_browser, temp_dir):
+    """
+    Configures and launches Firefox to run the vectorizer with.
+
+    Sets headless mode, sets the download directory to the desired output
+    directory, turns off page caching, and installs FathomFox and Fathom
+    Trainees.
+
+    We return the webdriver object, and the process IDs for both the Firefox
+    process and the geckoview driver process so we can shutdown either
+    gracefully or ungracefully.
+    """
+    print('Configuring Firefox...', end='', flush=True)
+    options = webdriver.FirefoxOptions()
+    options.headless = not show_browser
+
+    profile = webdriver.FirefoxProfile()
+    profile.set_preference('browser.download.folderList', 2)
+    profile.set_preference('browser.download.dir', str(pathlib.Path(output_directory).absolute()))
+    profile.set_preference('browser.cache.disk.enable', False)
+    profile.set_preference('browser.cache.memory.enable', False)
+    profile.set_preference('browser.cache.offline.enable', False)
+
+    firefox = webdriver.Firefox(
+        options=options,
+        firefox_profile=profile,
+        service_log_path=f'{temp_dir.name}/geckodriver.log',
+    )
+
+    firefox.install_addon(fathom_fox, temporary=True)
+    firefox.install_addon(fathom_trainees, temporary=True)
+    print('Done')
+    return firefox, firefox.capabilities['moz:processID'], firefox.service.process.pid
+
+
+def run_vectorizer(firefox, fathom_type, sample_filenames):
+    """
+    Set up the vectorizer and run it, creating the vectors file.
+
+    Navigate to the vectorizer page of FathomFox, paste the sample filenames
+    into the text area and hit the vectorize button.
+
+    We monitor the status text area for errors and to see how many samples
+    have been vectorized, so we know when the vectorizer has stopped running.
+    """
+    print('Configuring Vectorizer...', end='', flush=True)
+    # Navigate to the vectorizer page
+    fathom_fox_uuid = get_fathom_fox_uuid(firefox)
+    firefox.get(f'moz-extension://{fathom_fox_uuid}/pages/vector.html')
+
+    pages_text_area = firefox.find_element_by_id('pages')
+    pages_text_area.send_keys(sample_filenames)
+
+    ruleset_dropdown_selector = Select(firefox.find_element_by_id('ruleset'))
+    ruleset_dropdown_selector.select_by_visible_text(fathom_type)
+
+    downloads_dir = pathlib.Path(firefox.profile.default_preferences['browser.download.dir'])
+    vector_files_before = set(downloads_dir.glob('vector*.json'))
+    number_of_samples = len(sample_filenames.splitlines())
+    status_box = firefox.find_element_by_id('status')
+    vectorize_button = firefox.find_element_by_id('freeze')
+    completed_samples = 0
+    print('Done')
+
+    with progressbar(length=number_of_samples, label='Running Vectorizer...') as bar:
+        vectorize_button.click()
```

Comment:
When I run the Vectorizer with `-s`, sometimes the pages give the error that the server is taking too long to respond, and the connection timed out. Not all the time. I just ran this command `fathom-vectorize -s trainees.js article samples/training ~/Projects/fathom-fox/ ~/Projects/fathom-trainees/` twice, and the first time, it timed out with those errors, and the second time it vectorized successfully.

When it times out, I get the error: `Error: Invalid tab ID: X.` on the Vectorizer page.

Is it possible that the server sometimes drops some requests if they occur too quickly? This (particularly headless mode) is certainly the fastest we have ever requested these resources before from our HTTP server.

Perhaps we could file a follow up issue to investigate this more.

Reply:
Hmm, sadly, I haven't run into this. That could be. If you edit FathomFox to reduce the number of tabs (and reduce the speed of the requests), does the error go away?

Reply:
I just want to mention that I have since encountered this problem outside of this PR with the latest FathomFox with a parallelized Vectorizer, so it’s definitely not related to the work here.

After trying your suggestion, I think there are two different bugs here: "Unable to connect" (an error page that Firefox displays in the tab where the page would have loaded that does not stop the Vectorizer) and "Error: Invalid tab ID: X" (an error that the extension throws that shows up in the status area of the Vectorizer page, and does stop the Vectorizer).

I modified FathomFox to use `maxTabs` for 8, 16 (default) and 24 leaving the default delay of 5s unchanged; here are the number of failures for each error:

|`maxTabs` | No issues  | Unable to connect | Invalid tab ID |
| ------------- | ------------- | ------------- | ------------- |
| 8 | 3  | 0  |  2 |
| 16 | 2  | 2  |  1 |
| 24 | 0  | 5  |  0 |

So thanks to your suggestion, I do think that my hunch was correct, and we may want to revisit our `maxTabs` value. I will file two separate issues for these.